openhands/agenthub/codeact_agent/codeact_agent.py

```diff
@@ -286,8 +242,43 @@ def _get_messages(self, state: State) -> list[Message]:
             ),
             None,
         )
+        messages = (
+            [
+                Message(
+                    role='system',
+                    content=[
+                        TextContent(
+                            text=self.prompt_manager.get_system_message(),
+                            cache_prompt=self.llm.is_caching_prompt_active(),  # Cache system prompt
+                        )
+                    ],
+                ),
+                Message(
+                    role='user',
+                    content=[
+                        TextContent(
+                            text=self.prompt_manager.get_example_user_message(
+                                ''
+                                if latest_user_message is None
+                                else latest_user_message.content[-1].text
+                            ),
+                            cache_prompt=self.llm.is_caching_prompt_active(),  # if the user asks the same query,
+                        )
+                    ],
+                ),
+            ]
+            + messages
+        )
+
         if latest_user_message:
             reminder_text = f'\n\nENVIRONMENT REMINDER: You have {state.max_iterations - state.iteration} turns left to complete the task. When finished reply with <finish></finish>.'
             latest_user_message.content.append(TextContent(text=reminder_text))
 
+        if self.llm.is_caching_prompt_active():
+            user_turns_processed = 0
+            for message in reversed(messages):
+                if message.role == 'user' and user_turns_processed < 2:
+                    message.content[-1].cache_prompt = True
```

Comment:
It looks like the re-ordering here would make the reminder TextContent get the "cache_prompt"? That would break prompt caching, since the reminder text changes with each request so the cache is never hit.

Reply:
I'm pretty sure the logic is unchanged here, but could be wrong.

`messages` should be exactly the same here `(system, user example, rest)` as it was previously when we did this check. I just had to move the logic around, because now the `example_user_message` function needs the most recent user message in order to do the keyword-matching.

Reply:
LMK if there's a good way to test this

Reply:
Let's say the last user message is simple, no images, just a string:
- it will be in the Message, represented as a list with a single TextContent object
- we add the cache marker on it
- then we add the reminder as a second TextContent in the list
Result: ask the Anthropic API to cache the actual content only, not the reminder.

Currently on branch, if GitHub diff doesn't fool me 😅 
- the last Message would have in its list the TextContent with the actual content
- we add the reminder in a second TextContent
- we add cache marker to the last TextContent => that's the reminder now.

Reply:
ohhhh I see what you're saying :/

Let me bring this up in slack

Reply:
Re: testing. It's Anthropic only, and we have to send a few more messages than the first two. The log in console should show both cache writes and cache hits.

We might end up seeing that after some 4k (system message, user message), it does or doesn't hit [any more tokens](https://github.com/All-Hands-AI/OpenHands/pull/3773#issuecomment-2339022927)

Reply:
OK I think this is better now. We will deterministically inject the microagent text into any user message, and preserve those injections on subsequent runs, even if a new user message comes in